aesd-char-driver/main.c

```diff
@@ -184,13 +185,118 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
     mutex_unlock(&aesd_dev->aesd_mutex);
     kfree(write_buf);
 
+    // advance the pointer by the number of bytes written
+    *f_pos += count;
+
     return count;
 }
 
+loff_t aesd_llseek(struct file *filp, loff_t offset, int whence)
+{
+    struct aesd_dev *aesd_dev = NULL;
+    uint8_t index = 0;
+    struct aesd_buffer_entry *entry;
+    loff_t total_size = 0;
+    loff_t ret = 0;
+
+    if(filp == NULL)
+    {
+        return -EINVAL;
+    }
+
+    aesd_dev = filp->private_data;
+
+    // Lock mutex before acessing the global data
+    if (mutex_lock_interruptible(&aesd_dev->aesd_mutex))
+    {
+        PDEBUG("mutex_lock_interruptible failed");
+        return -ERESTARTSYS;
+    }
+
+    // Calculate the total size of all contents of the circular buffer
+    AESD_CIRCULAR_BUFFER_FOREACH(entry,&aesd_dev->buffer,index) {
+        total_size += entry->size;
+    }
+
+    ret = fixed_size_llseek(filp, offset, whence, total_size);
```

Comment:
@krishnasuhagiya29 no need to lock here through the end of the function since `aesd_dev` is not accessed.